Support for multi tenancy column discriminator (#32855)
quarkusio/quarkus: docs/src/main/asciidoc/hibernate-orm.adoc

"The term multitenancy, in general, is applied to software development to indicate an architecture in which a single running instance of an application simultaneously serves multiple clients (tenants). This is highly common in SaaS solutions. Isolating information (data, customizations, etc.) pertaining to the various tenants is a particular challenge in these systems. This includes the data owned by each tenant stored in the database" (link:{orm-doc-url-prefix}#multitenacy[Hibernate User Guide]).

> Quarkus currently supports the link:{orm-doc-url-prefix}#multitenacy-separate-database[separate database], the link:{orm-doc-url-prefix}#multitenacy-separate-schema[separate schema] approach and the link:{orm-doc-url-prefix}#multitenacy-discriminator[discriminator] approach.

Reviewer: ```suggestion
Quarkus currently supports the link:{orm-doc-url-prefix}#multitenacy-separate-database[separate database] approach, the link:{orm-doc-url-prefix}#multitenacy-separate-schema[separate schema] approach and the link:{orm-doc-url-prefix}#multitenacy-discriminator[discriminator] approach.
```

INSERT INTO known_fruits(id, name) VALUES (3, 'Blackberries');
----

==== DISCRIMINATOR approach

> The default data source will be used for all tenants. All entities defining a field annotated with `@TenandId` will be handle automatically.

Reviewer: ```suggestion
The default data source will be used for all tenants. All entities defining a field annotated with `@TenandId` will have that field populated automatically, and will get filtered automatically in queries.
```

[source,properties]
----
# Disable generation
quarkus.hibernate-orm.database.generation=none

> # Enable DISCRIMINATOR approach and use default datasource

Reviewer: ```suggestion
# Enable DISCRIMINATOR approach
```

quarkus.hibernate-orm.multitenant=DISCRIMINATOR

# The default data source used for all tenant schemas
quarkus.datasource.db-kind=postgresql
quarkus.datasource.username=quarkus_test
quarkus.datasource.password=quarkus_test
quarkus.datasource.jdbc.url=jdbc:postgresql://localhost:5432/quarkus_test

# Enable Flyway configuration
quarkus.flyway.migrate-at-start=true
----

> Here is an example of the Flyway SQL (`V1.0.0__create_fruits.sql`) to be created in the configured folder `src/main/resources/schema`.
> 
> [source,sql]
> ----
> CREATE TABLE known_fruits
> (
>   tenantId  VARCHAR(40),
>   id        INT,
>   name      VARCHAR(40)
> );
> CREATE SEQUENCE known_fruits_id_seq START WITH 7;
> INSERT INTO known_fruits(tenantId, id, name) VALUES ('base', 1, 'Cherry');
> INSERT INTO known_fruits(tenantId, id, name) VALUES ('base', 2, 'Apple');
> INSERT INTO known_fruits(tenantId, id, name) VALUES ('base', 3, 'Banana');
> 
> INSERT INTO known_fruits(tenantId, id, name) VALUES ('mycompany', 4, 'Avocado');
> INSERT INTO known_fruits(tenantId, id, name) VALUES ('mycompany', 5, 'Apricots');
> INSERT INTO known_fruits(tenantId, id, name) VALUES ('mycompany', 6, 'Blackberries');
> ----
> 
> 

Reviewer: I would remove this as Flyway is not strictly necessary with this approach.

```suggestion
```
